Add power management support for MXRT595

mmahadevan108 opened the pull request:
Power modes supported  are 	PM_STATE_RUNTIME_IDLE, PM_STATE_SUSPEND_TO_IDLE, PM_STATE_SOFT_OFF

mmahadevan108 committed:
soc: nxp: Add Power Management support for RT5xx

1. Support Sleep, Deep Sleep and Deep Power down modes
2. Enable the MEMC FlexSPI driver when using device power
   management so we can reconfigure the FlexSPI pins to
   save power. The MEMC FlexSPI driver is enabled when we
   enable the Flash subsystem, however we would like to
   reconfigure the FlexSPI pins even when the Flash driver
   is disabled, hence MEMC is selected when PM_DEVICE
   is turned on.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
boards: rt595: Changes to support Power management

1. Add code to setup the PMIC voltages during low
   power modes
2. Add Zephyr power states that are supported to
   device tree
3. Add low power pin configration for Flash pins
   to pinctrl

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
tests: pm_mgmt_soc: Enable additional configs for RT595 EVK

The FlexSPI pins are reconfigured before going into Deep
Sleep mode to save power. The requires enabling
CONFIG_PM_DEVICE so we can access the flexspi pin settings.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
tests: power_mgmt_soc: Add overlay for MXRT595 EVK

Enable the OS Timer to be a wakeup source

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
drivers: counter: Update NXP LPC RTC for wakeup source

Add support to wakeup from low power modes

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
dts: nxp: power: Add a property for low power mode behavior

Add a property to configure blocks that should be enabled
during deep sleep mode

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

DerekSnell reviewed (dismissed):
Great job, @mmahadevan108 

I have tested this on my EVK and see some great power savings.  Overall this PR looks really good to me, but I have a few non-blocking comments for you.

Thanks for all your efforts with this.

DerekSnell commented:
@mmahadevan108 ,
Something else I just remembered, please update the [board feature table](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/mimxrt595_evk/doc/index.rst#supported-features) with a row for Power Manager.  Similar to this [other board's feature table](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/npcx9m6f_evb/doc/index.rst#supported-features).

Thanks

ceolin commented on boards/arm/mimxrt595_evk/board.c:
```diff
@@ -4,8 +4,178 @@
+	 */
+
+	/* SW1 setting in Run mode: Enabled @1V */
+	ret = regulator_set_voltage(sw1, 1000000, 1000000);
```
wouldn't be better having these voltage values defined in DT ?

mmahadevan108 replied:
The voltage levels are pre-defined in the SoC datasheet. I do not see a need for the user to specify these values.

mmahadevan108 replied:
@ceolin after discussing with the internal team it seem like having this come from DT is a good idea. 
Do you have a suggestion on where to place these PMIC properties?

mmahadevan108 replied:
I have updated the implementation to use DT

ceolin commented on soc/arm/nxp_imx/rt5xx/power.c:
```diff
@@ -0,0 +1,163 @@
+		POWER_EnterDeepSleep(APP_EXCLUDE_FROM_DEEPSLEEP);
+		restore_deepsleep_pin_config();
+		break;
+	case PM_STATE_SOFT_OFF:
```
Is this state defined in DT ? I just saw runtime-idle and suspend-to-idle.

mmahadevan108 replied:
It is not. For putting this in DT I will have to add dummy residency values as we not enter these states from the idle loop.
This state should be entered only when the user application decides to put the platform in SOFT_OFF, recovery from this state causes a `warm-reset`.
This is also similar to other implementations of `PM_STATE_SOFT_OFF`

mmahadevan108 commented on soc/arm/nxp_imx/rt5xx/power.c:
```diff
@@ -0,0 +1,163 @@
+
+/*!< Power down all unnecessary blocks */
+#define NODE_ID DT_INST(0, nxp_pdcfg_power)
+#define APP_EXCLUDE_FROM_DEEPSLEEP ((const uint32_t[]) \
```
Fixed, thanks for spotting this.

mmahadevan108 commented on boards/arm/mimxrt595_evk/mimxrt595_evk_cm33.dts:
```diff
@@ -345,6 +374,13 @@ zephyr_udc0: &usbhs {
 	status = "okay";
+	pinctrl-0 = <&pinmux_flexspi>;
+	pinctrl-1 = <&pinmux_flexspi_sleep>;
+#if (CONFIG_XIP && CONFIG_PM)
```
Thank you Daniel, I have fixed this.

mmahadevan108 commented on soc/arm/nxp_imx/rt5xx/power.c:
```diff
@@ -0,0 +1,163 @@
+	isp_pin[2] = IOPCTL->PIO[3][29];
+
+	/* Disable ISP Pin pull-ups and input buffers to avoid current leakage */
+	IOPCTL->PIO[1][15] = 0;
```
I am not clear where these ISP pins would fit in device tree. These pins decide the boot flow and are used by the ROM.

danieldegrasse replied:
Given that we are restoring the pin setting after LPM exit, and these pins are configured as pull-up/input buffer enable at boot I am ok with leaving this configuration as is.

mmahadevan108 commented on boards/arm/mimxrt595_evk/mimxrt595_evk_cm33.dts:
```diff
@@ -92,6 +92,38 @@
+			power-state-name = "suspend-to-idle";
+			min-residency-us = <500>;
+			exit-latency-us = <120>;
+			deep-sleep-config = <0xC800>,
```
I have added a comment block. There is also some information in the bindings file.

mmahadevan108 committed:
samples: mimxrt595: Add Deep power down sample

Add a sample to put the part into Deep power down mode by
setting the state to PM_STATE_SOFT_OFF.
Wakup the part using an RTC alarm.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 commented:
> @mmahadevan108 , Something else I just remembered, please update the [board feature table](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/mimxrt595_evk/doc/index.rst#supported-features) with a row for Power Manager. Similar to this [other board's feature table](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/npcx9m6f_evb/doc/index.rst#supported-features).
> 
> Thanks

Thanks, I have added this information

danieldegrasse commented on dts/bindings/regulator/regulator-pmic.yaml:
```diff
@@ -125,3 +125,23 @@ child-binding:
+      regulator-microvolt:
+        type: int
+        required: false
+        description: the default voltage which regulator would be set in.
```
Could you add to this description, clarifying that this voltage is used for a given power mode?

mmahadevan108 replied:
Done

danieldegrasse commented on soc/arm/nxp_imx/rt5xx/power.c:
```diff
@@ -0,0 +1,223 @@
+/* System clock frequency. */
+extern uint32_t SystemCoreClock;
+
+#define FLEXSPI_DEVICE DT_NODELABEL(flexspi)
```
I think for these lines you may want to also check that pin control nodes are present for the FlexSPI device. Otherwise I believe that `PINCTRL_DT_DEFINE()` may result in a build error.  `DT_NODE_HAS_PROP(FLEXSPI_DEVICE, pinctrl)` should work for this

mmahadevan108 replied:
i have tried by deleting the `pinctrl` properties deleted and do not see any build failures.

danieldegrasse commented on tests/subsys/pm/power_mgmt_soc/boards/mimxrt595_evk_cm33.conf:
```diff
@@ -0,0 +1,11 @@
+# Turn on Device Level Power Management as we wish
+# to reconfigure the FlexSPI pins for power savings
+# when transitioning the SoC to Deep Low Power modes.
+CONFIG_PM_DEVICE=y
```
Why is `CONFIG_PM_DEVICE` required here, but not in the board specific system off sample?

mmahadevan108 replied:
Thanks for catching this, I have added to the sample as well.

gmarull commented on dts/bindings/regulator/regulator-pmic.yaml:
```diff
@@ -125,3 +125,23 @@ child-binding:
+      Sub-nodes for the regulator state in different Zephyr Power management states.
+
+    properties:
+      regulator-on:
+        type: boolean
+        required: false
+        description: regulator should be on.
+
+      regulator-off:
+        type: boolean
+        required: false
+        description: regulator should be off.
+
+      regulator-microvolt:
+        type: int
+        required: false
+        description: the default voltage which regulator would be set in.
```
please, let's stick to Linux standard properties (e.g. `regulator-boot-on`). See related work here https://github.com/zephyrproject-rtos/zephyr/pull/52326

mmahadevan108 replied:
These are similar to the Linux properties, however I dropped the `suspend` as these properties apply to the Zephyr `PM_STATE_ACTIVE` state as well.

mmahadevan108 replied:
@gmarull, sounds ok to you?

gmarull replied:
we should be following this: https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/regulator/regulator.yaml#L286. This is something that the PM subsystem should actually facilitate, but in absence of proper infrastructure, I guess we'll need to wait. However, nothing prevents us to at least follow the same bindings schema.

mmahadevan108 replied:
@gmarull. I have rebased and updated to use the latest changes made to the regulator driver. I have also changed the property names to match Linux

mmahadevan108 replied:
@gmarull I have rebased and updated the code base to use the latest changes made to the regulator API. Based on the new design, I have dropped the new properties added. Kindly take a look and let me know if you see any other issues.

DerekSnell commented on soc/arm/nxp_imx/rt5xx/power.c:
```diff
@@ -0,0 +1,223 @@
+	}
+#endif
+
+	POWER_SetLdoVoltageForFreq(SystemCoreClock, 0);
```
This API configures the RT500's internal LDO, not used with the PMIC.  Instead call BOARD_SetPmicVoltageForFreq()

mmahadevan108 replied:
I have added a comment on why this API is called.

gmarull commented on soc/arm/nxp_imx/rt5xx/power.c:
```diff
@@ -0,0 +1,226 @@
+#define FLEXSPI2_DEVICE DT_NODELABEL(flexspi2)
+#if DT_NODE_HAS_STATUS(FLEXSPI2_DEVICE, okay)
+
+/*
+ * Declare pin configuration state for flexspi2 pins.
+ * These pins will be reconfigured for power savings
+ */
+PINCTRL_DT_DEFINE(FLEXSPI2_DEVICE);
+const struct pinctrl_dev_config *flexspi2_pincfg = PINCTRL_DT_DEV_CONFIG_GET(FLEXSPI2_DEVICE);
+const struct pinctrl_state *flexspi2_pins_default_state;
+const struct pinctrl_state *flexspi2_pins_sleep_state;
+/* Flag indicates if we can reconfigure flexspi pins for additional power savings */
+static bool flexspi2_pin_reconfig;
```
this needs to be managed by the device, not outside. Don't you have a driver for flexspi2?

mmahadevan108 replied:
These need to be set close to entering Low Power modes. Hence they are done here.

gmarull replied:
The problem with this approach is that it will conflict with e.g. dynamic pinctrl (you'll end up with duplicate symbols). What I'd suggest is to create a new compatible where to store such description (maybe re-use nxp,pdcfg-power by adding there pinctrl-N entries). Also, please document that this has to be done independently of the flash driver to save power in all configurations.

mmahadevan108 replied:
@gmarull, thank you for your comments. Turns out we can enable the FlexSPI driver without having to pull in the whole Flash subsystem.
I have moved this code to the FlexSPI driver by adding Device Level Power Management. This did uncover an issue in the Zephyr PM code. I will submit a separate PR for that.

mmahadevan108 replied:
@gmarull, let me know if you have any more comments. We would like to get this is in time for the next release.

gmarull commented on boards/arm/mimxrt595_evk/board.c:
```diff
@@ -71,6 +72,11 @@ static int board_config_pmic(const struct device *dev)
 		return ret;
 	}
 
+#if CONFIG_PM
+	/* We can enter deep low power modes */
+	pm_policy_state_lock_put(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+#endif
```
ifdeffery not needed, stubs are provided.

mmahadevan108 replied:
Thanks, updated.

mmahadevan108 committed:
drivers: memc: Add device PM to MCUX Flexspi driver

Add Device level power manangement support to the
MCUX FlexSPI driver.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

DerekSnell reviewed (commented):
LGTM, but waiting on others to review

dleach02 commented:
@gmarull can you revisit this PR to determine if your change requests have been met? We need to get this PR completed before the release window becomes a problem.

gmarull commented on boards/arm/mimxrt595_evk/mimxrt595_evk_cm33.dts:
```diff
@@ -91,6 +91,48 @@
+		 */
+	};
+
+	flexspi_lowpower: flexspi_lowpower {
+		compatible = "nxp,flexspi-lowpower";
+		pinctrl-0 = <&pinmux_flexspi>;
+		pinctrl-1 = <&pinmux_flexspi_sleep>;
+		pinctrl-names = "default", "sleep";
+	};
```
where is this used now?
